solutions/platinum/ioi-05-mountain.mdx

```diff
@@ -0,0 +1,179 @@
+---
+id: ioi-05-mountain
+source: IOI
+title: 2005 - Mountain
+author: Justin Ji
+---
+
+[Official Analysis](https://dmoj.ca/problem/ioi05p3/editorial)
+
+Consider an easier version of the problem where $N \leq 10^{6}$. In that case,
+we use a [maximum prefix sum segment tree](https://usaco.guide/problems/cses-2166-prefix-sum-queries/solution)
+with lazy updates. Then, to find the last index where the maximum prefix sum
+$\leq H$, we walk on the segment tree.
+
+However, $N \leq 10^{9}$, rendering the approach above invalid. The approach
+I will describe below does not utilize sparse segment tree.
+
+Given that we do not need to answer the queries online, we can coordinate compress
+the values. Then, each index on the segment tree represents a range, instead
+of being an individual location. Let's say we have range updates on the
+following subarrays:
+
+$$[1, 10], [2, 5]$$
+
+Then, we can maintain the following distinct values:
+
+$$[1, 2, 6, 11]$$
+
+These distinct values define the ranges $$[1, 2], [2, 5], [6, 10]$$. Then, if we
+want to update the range $$[1, 10]$$, we can update the three subranges listed
+above. The idea is that if we maintain an array $vals[i]$, then each index $i$
+will represent the range $$[vals[i], vals[i + 1] - 1]$$. For each update on
+the range $$[a, b]$$, the array $vals$ should contain $a$ and $b + 1$.
+
+After figuring out the coordinate compression, most of the segment tree is
+pretty similar to the approach described in the beginning.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(Q\log{Q})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+class LazySegtree {
```

Comment:
how different is this from a stock lazy segtree?

Reply:
it practically is a stock lazy segtree, technically the segtree walking is slightly diff but I think its fine

Reply:
does it affect the logic in any way
can i shove it in a cses problem and have it work as expected

Reply:
with a bit of modification, probably

[idk, I use this most of the time](https://github.com/TheGamingMousse/use-a-cow/blob/main/Implementation/SegtreeRecursive.cpp)

Reply:
uh mind telling me why this one is different

Reply:
segtree walking is different here because we need to maintain the result for our current prefix at every step, which is typically unnecessary. and also, instead of just returning the indice, we need to do some math to get the answer at the end

Reply:
can you mark where it differs with comments then

Reply:
sure

Reply:
done